tests/std/tests/VSO_0157762_feature_test_macros/test.cpp

```diff
@@ -5,465 +5,431 @@
 
 int main() {} // COMPILE-ONLY
 
-// COMPILER FEATURE-TEST MACROS
+// ATTRIBUTE FEATURE-TEST MACROS.
 
-#ifdef _MSVC_LANG
-#if _MSVC_LANG > 201703L
-#define CXX17_MODE 1
-#define CXX20_MODE 1
-#elif _MSVC_LANG > 201402L
-#define CXX17_MODE 1
-#define CXX20_MODE 0
+#ifdef __has_cpp_attribute
+// Good.
 #else
-#define CXX17_MODE 0
-#define CXX20_MODE 0
+#error Expected __has_cpp_attribute to be defined.
 #endif
+#if defined(__has_cpp_attribute)
+// Good.
 #else
-#if __cplusplus > 201703L
-#define CXX17_MODE 1
-#define CXX20_MODE 1
-#elif __cplusplus > 201402L
-#define CXX17_MODE 1
-#define CXX20_MODE 0
-#else
-#define CXX17_MODE 0
-#define CXX20_MODE 0
+#error Expected __has_cpp_attribute to be defined.
 #endif
+#ifndef __has_cpp_attribute
+#error Expected __has_cpp_attribute to be defined.
 #endif
 
+#if __has_cpp_attribute(carries_dependency) != 200809L
+#error Expected has_cpp_attribute(carries_dependency) to equal 200809L.
+#endif
 
-// Always defined to varying values (C++14 versus C++17-and-newer mode).
+#if __has_cpp_attribute(deprecated) != 201309L
+#error Expected has_cpp_attribute(deprecated) to equal 201309L.
+#endif
 
-#ifndef __cpp_constexpr
-#error BOOM
-#elif CXX17_MODE
-#if __cpp_constexpr != 201603L
-#error BOOM
-#else
-STATIC_ASSERT(__cpp_constexpr == 201603L);
+#if __has_cpp_attribute(noreturn) != 200809L
+#error Expected has_cpp_attribute(noreturn) to equal 200809L.
+#endif
+
+#if defined(__clang__) || defined(__EDG__) // clang and EDG don't yet implement P1771R1
+#if __has_cpp_attribute(nodiscard) != 201603L
+#error Expected has_cpp_attribute(nodiscard) to equal 201603L.
 #endif
 #else
-#if __cpp_constexpr != 201304L
-#error BOOM
-#else
-STATIC_ASSERT(__cpp_constexpr == 201304L);
+#if __has_cpp_attribute(nodiscard) != 201907L
+#error Expected has_cpp_attribute(nodiscard) to equal 201907L.
 #endif
 #endif
 
-#ifndef __cpp_inheriting_constructors
-#error BOOM
-#elif (CXX17_MODE || defined(__clang__)) && !defined(__EDG__) // Clang implemented this C++17 feature unconditionally.
-                                                              // TRANSITION, VSO-610203
-#if __cpp_inheriting_constructors != 201511L
-#error BOOM
-#else
-STATIC_ASSERT(__cpp_inheriting_constructors == 201511L);
+#if _HAS_CXX17 || defined(__clang__) || defined(__EDG__) // Clang and EDG provide this in C++14 mode.
+#if __has_cpp_attribute(fallthrough) != 201603L

```

Comment:
Ditto "fallthrough isn't in the order I expect from the issue discussion." (I'll stop commenting on ordering now - there are a lot of these that don't match my expectations.)